group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/TopicMetadata.java

```diff
@@ -40,23 +45,31 @@ public class TopicMetadata {
      */
     private final int numPartitions;
 
+    /**
+     * Map of every partition to a set of its rackIds.
+     * If the rack information is unavailable, this is an empty map.
+     */
+    private final Map<Integer, Set<String>> partitionRacks;
+
     public TopicMetadata(
         Uuid id,
         String name,
-        int numPartitions
+        int numPartitions,
+        Map<Integer, Set<String>> partitionRacks
     ) {
-        this.id = Objects.requireNonNull(id);
-        if (Uuid.ZERO_UUID.equals(id)) {
-            throw new IllegalArgumentException("Topic id cannot be ZERO_UUID.");
-        }
-        this.name = Objects.requireNonNull(name);
-        if (name.isEmpty()) {
-            throw new IllegalArgumentException("Topic name cannot be empty.");
-        }
-        this.numPartitions = numPartitions;
-        if (numPartitions < 0) {
-            throw new IllegalArgumentException("Number of partitions cannot be negative.");
-        }
+            this.id = Objects.requireNonNull(id);
+            this.partitionRacks = Objects.requireNonNull(partitionRacks);
```

Comment:
nit: Could we put this one last in the constructor? The line before it and after it goes together.